src/modm/math/geometry/vector.hpp

```diff
@@ -11,141 +12,518 @@
+	 * toOrthogonalVector() method, which returns a perpendicular copy
+	 * of the vector.
 	 */
-	template<typename T, uint8_t N>
-	class Vector
+	WideType operator^ (const Vector &other) const requires (N == 2)
+	{ return WideType(x()) * WideType(other.y()) - WideType(y()) * WideType(other.x()); }
+
+	Vector operator^ (const Vector &other) const requires (N == 3)
 	{
-	public:
-		Vector();
-		Vector(const T *ptData);
-
-		Vector(const Matrix<T, N, 1> &rhs);
-		Vector& operator = (const Matrix<T, N, 1> &rhs);
-
-		bool operator == (const Vector &rhs) const;
-		bool operator != (const Vector &rhs) const;
-		bool operator < (const Vector &rhs) const;
-		bool operator <= (const Vector &rhs) const;
-		bool operator > (const Vector &rhs) const;
-		bool operator >= (const Vector &rhs) const;
-
-		const T& operator [] (uint8_t index) const;
-		T& operator [] (uint8_t index);
-
-		T* ptr();
-		const T* ptr() const;
-
-		Vector operator + (const Vector &rhs) const;
-		Vector operator - (const Vector &rhs) const;
-		T operator * (const Vector &rhs) const;
-		Vector operator * (const T &rhs) const;
-		Vector operator / (const T &rhs) const;
-		Vector& operator += (const Vector &rhs);
-		Vector& operator -= (const Vector &rhs);
-		Vector& operator *= (const T &rhs);
-		Vector& operator /= (const T &rhs);
-		Vector& operator - ();		// FIXME
-
-		T getLength() const;
-		T getLengthSquared() const;
-
-		Matrix<T, N, 1>&
-		asMatrix();
-
-		const Matrix<T, N, 1>&
-		asMatrix() const;
-
-		Matrix<T, 1, N>&
-		asTransposedMatrix();
-
-		const Matrix<T, 1, N>&
-		asTransposedMatrix() const;
-
-	public:
-		static inline uint8_t
-		getSize();
-
-		T coords[N];
-	};
-
-	template< typename T, uint8_t N >
-	struct detail::MakeSigned< Vector<T, N> >
-	{ using type = Vector< SignedType<T>, N >; };
-
-	template< typename T, uint8_t N >
-	struct detail::MakeUnsigned< Vector<T, N> >
-	{ using type = Vector< UnsignedType<T>, N >; };
-
-	template< typename T, uint8_t N >
-	struct detail::WideType< Vector<T, N> >
-	{ using type = Vector< WideType<T>, N >; };
-}
+		return Vector(
+			y() * other.z() - z() * other.y(),
+			z() * other.x() - x() * other.z(),
+			x() * other.y() - y() * other.x()
+		);
+	}
+
+	[[deprecated("Use Vector<T, N>::operator^() instead!")]]
+	constexpr WideType cross(const Vector& other) const
+	{ return operator^(other); }
+	// auto cross = operator^; // Why is this 'function alias' not accepted?
+
+	// various methods for the lengh of the vector
+
+	constexpr auto getLengthSquared() const {
+		// TODO treat the right CalcType
+		// - is always unsigned
+		// - Must fit std::pow(std::numeric_limits<T>::max(), 2) * N
+		using CalcType = WideType;
+
+		CalcType sum{0};
+
+		for(auto &v: *this)
+			sum += WideType(v) * WideType(v);
+		return sum;
+	}
+
+	template<typename TR = T>
+	constexpr TR getLength() const
+	{ return modm::round_smart<TR, decltype(std::sqrt(getLengthSquared()))>(std::sqrt(getLengthSquared())); }
```

Comment:
No need for the `decltype`, the second parameter will be deduced from the argument. That is a lot cleaner.
```suggestion
	{ return modm::round_smart<TR>(std::sqrt(getLengthSquared())); }
```